### PR TITLE
[SPARK-7938][BUILD]Use Google ErrorProne during Maven build of Spark

### PR DESCRIPTION
**Pull request by yjshen**
[Error Prone](http://errorprone.info) augment the compiler’s type analysis and hooks into standard build, aiming at detecting Java code mistakes immediately after they are made.
This PR hooks ErrorProne into maven-compiler-plugin to enable code analysis at maven's compile & test-compile phases.

[Review on Reviewable](https://reviewable.io/reviews/apache/spark/6515)

### CODE REVIEW

**Comment by rxin** [comment]
Thanks, @yijieshen. Have you tried running this locally and see what was produced?

**Comment by rxin** [comment]
Jenkins, ok to test.

**Comment by yijieshen** [comment]
@rxin, I've tried locally, when compiler's warning are suppressed by default, nothing error occurs. When I turned on the warnings, it only warns on spurious wakeups of a test class, I would paste them here later.

**Comment by yijieshen** [comment]
@rxin: 
/Users/yijie/code/spark/network/common/src/test/java/org/apache/spark/network/RequestTimeoutIntegrationSuite.java:114: warning: [WaitNotInLoop] Object.wait() should always be called in a loop
      callback0.wait(FOREVER);
                    ^
    (see http://errorprone.info/bugpattern/WaitNotInLoop)
/Users/yijie/code/spark/network/common/src/test/java/org/apache/spark/network/RequestTimeoutIntegrationSuite.java:122: warning: [WaitNotInLoop] Object.wait() should always be called in a loop
      callback1.wait(4 * 1000);
                    ^
    (see http://errorprone.info/bugpattern/WaitNotInLoop)
/Users/yijie/code/spark/network/common/src/test/java/org/apache/spark/network/RequestTimeoutIntegrationSuite.java:162: warning: [WaitNotInLoop] Object.wait() should always be called in a loop
      callback0.wait(FOREVER);
                    ^
    (see http://errorprone.info/bugpattern/WaitNotInLoop)
/Users/yijie/code/spark/network/common/src/test/java/org/apache/spark/network/RequestTimeoutIntegrationSuite.java:174: warning: [WaitNotInLoop] Object.wait() should always be called in a loop
      callback1.wait(FOREVER);
                    ^
    (see http://errorprone.info/bugpattern/WaitNotInLoop)
/Users/yijie/code/spark/network/common/src/test/java/org/apache/spark/network/RequestTimeoutIntegrationSuite.java:222: warning: [WaitNotInLoop] Object.wait() should always be called in a loop
      callback0.wait(2 * 1000);
                    ^
    (see http://errorprone.info/bugpattern/WaitNotInLoop)
/Users/yijie/code/spark/network/common/src/test/java/org/apache/spark/network/sasl/SparkSaslSuite.java:277: warning: [WaitNotInLoop] Object.wait() should always be called in a loop
        lock.wait(10 * 1000);
                 ^
    (see http://errorprone.info/bugpattern/WaitNotInLoop)

**Comment by rxin** [comment]
Thanks. Can you manually inject a "bug" into some Java code to see if it reports that correctly?

If those are all the warnings, I'd just update the code to remove the warnings, and then turn warnings on as well.

**Comment by yijieshen** [comment]
@rxin, I have insert [DeadException](http://errorprone.info/bugpattern/DeadException) `new Exception()`which is considered as ERROR in error prone, it reports correctly. Will test all the ERROR bug patterns manually to make sure.

**Comment by rxin** [comment]
No need to test all of them. As long as one works ... :)

**Comment by yijieshen** [comment]
Test more Error bug patterns manually, work as expected.
However, after an error report, if no `mvn clean` called before next `mvn package`, and not changing the buggy code, `maven-compiler-plugin` would just think all classes are up to date and not report the error again.

**Comment by rxin** [comment]
Got it - as long as it is fine on Jenkins, I think it is OK. We can rely on Jenkins to do this anyway.

One thing to add is lint-java file in /dev folder.

**Comment by yijieshen** [comment]
Not that sure how lint-java works. If I understand correctly, in lint-java, we should run error prone separately like what lint-scala does? 
If so, seems not possible this moment, since error prone can only tightly coupled with maven's `compile` and `test-compile`, refer to issue: https://github.com/google/error-prone/issues/337

**Comment by rxin** [comment]
Ah ok. So the only way to run this is to use mvn compile?

**Comment by yijieshen** [comment]
I think so, would keep an eye on this and create lint-java if possible later.

**Comment by rxin** [comment]
Jenkins, test this please.

File: pom.xml [hunk 1]
```diff
@@ -1226,7 +1226,16 @@
             <encoding>UTF-8</encoding>
             <maxmem>1024m</maxmem>
             <fork>true</fork>
+            <compilerId>javac-with-errorprone</compilerId>
```
Comment (srowen) [inline pom.xml]: Can this be put in a profile? I'm slightly concerned about pushing the entire build through a different compiler just for this tool. It's easy to make this selectable.

Comment (yijieshen) [inline pom.xml]: move it into a profile and turned on the compiler warnings since the profile itself is optional.

**Comment by yijieshen** [comment]
How can I ask Jenkins to retest it?

**Comment by rxin** [comment]
Jenkins, retest this please.

**Comment by rxin** [comment]
Jenkins, retest this please.

**Comment by rxin** [comment]
@yijieshen how do I run this?

**Comment by yijieshen** [comment]
@rxin, just add -Pep-tests when running mvn package
I've turned on warnings in the second commit since it's optional and I assume run this deliberately would love to see all the warnings

**Comment by rxin** [comment]
Thanks @yijieshen. Just tried this locally and worked. Can you change the report level from warning to error, and fix the few instances of warnings? The reason is I don't think it'd be very useful to run this at warning level, since there are so many warning messages that people will just ignore. The only way to make this useful is to fail the build if there are violations.

**Comment by JoshRosen** [comment]
Since this is now guarded behind a special Maven profile, we should also make sure to update the Jenkins build to run with this enabled.  AFAIK the Jenkins Maven builds don't actually run through run-tests-jenkins so we'll have to log in and make the changes via the Jenkins admin panel.

**Comment by yijieshen** [comment]
@rxin, in error prone, some checks are `on` by default, and others are experimental and `off` by default. Each of the checks is considered with a `Severity level`, some are `ERROR` such as DEADException, mean while others are considered `WARNING`. 
So you mean we could make all the `on` & `WARNING` checks as `on` & `ERROR` first, and fix them?

**Comment by yijieshen** [comment]
@JoshRosen, since the profile is tightly coupled with `compile` & `test-compile` goals in maven, is it OK to run `mvn package` in `bin/run-tests`? e.g. in `lint-java`?

**Comment by yijieshen** [comment]
@rxin, just turned on all the mature checkers from warning into error in the latest commit, implementing `lint` now.

**Comment by andrewor14** [comment]
retest this please

**Comment by andrewor14** [comment]
@harishreedharan @vanzin @srowen maybe the change from #6600 will fail here :)

**Comment by yijieshen** [comment]
@aarondav, I've moved the callback.wait into `while (callback0.success == null && callback0.failure == null) ` loops since it seems like just waiting for either success or failure, and in this way we could guard against spurious wakeups. Please check the commit above to see if I made mistakes.

**Comment by yijieshen** [comment]
@vanzin , it seems no variable to check [here](https://github.com/yijieshen/spark/blob/errorprone/network/common/src/test/java/org/apache/spark/network/sasl/SparkSaslSuite.java#L278-278), so I just suppress `WaitNotInLoop` Error, am I right?

**Comment by vanzin** [comment]
@yijieshen sounds right; it's a test anyway, so adding a variable to check there sounds like overkill.

**Comment by yijieshen** [comment]
@JoshRosen , I've made error prone check inside `dev/lint-java` and also put it into `dev/run-tests`.
Since error prone can only be hooked into `compile` and `test-compile` goals in Maven's lifecycle,  we need to run at least `mvn test-compile` to ensure it checks both src/main/java and src/test/java, so it may take more times than other `lint`, is this OK?

**Comment by yijieshen** [comment]
@rxin, can you help me to review the latest commit?

**Comment by yijieshen** [comment]
@srowen , can you help me to review the latest commit?

**Comment by yijieshen** [comment]
Not sure if it was too strict to level up `EqualsHashCode: Classes that override equals should also override hashCode.`  from warning to error.

/Users/yijie/code/spark/network/common/src/main/java/org/apache/spark/network/protocol/ChunkFetchSuccess.java:65: error: [EqualsHashCode] Classes that override equals should also override hashCode.
  public boolean equals(Object other) {
                 ^
    (see http://errorprone.info/bugpattern/EqualsHashCode)
/Users/yijie/code/spark/network/common/src/main/java/org/apache/spark/network/protocol/RpcResponse.java:54: error: [EqualsHashCode] Classes that override equals should also override hashCode.
  public boolean equals(Object other) {
                 ^
    (see http://errorprone.info/bugpattern/EqualsHashCode)
/Users/yijie/code/spark/network/common/src/main/java/org/apache/spark/network/protocol/RpcRequest.java:63: error: [EqualsHashCode] Classes that override equals should also override hashCode.
  public boolean equals(Object other) {
                 ^
    (see http://errorprone.info/bugpattern/EqualsHashCode)
/Users/yijie/code/spark/network/common/src/main/java/org/apache/spark/network/protocol/ChunkFetchRequest.java:52: error: [EqualsHashCode] Classes that override equals should also override hashCode.
  public boolean equals(Object other) {
                 ^
    (see http://errorprone.info/bugpattern/EqualsHashCode)
/Users/yijie/code/spark/network/common/src/main/java/org/apache/spark/network/protocol/RpcFailure.java:54: error: [EqualsHashCode] Classes that override equals should also override hashCode.
  public boolean equals(Object other) {
                 ^
    (see http://errorprone.info/bugpattern/EqualsHashCode)
/Users/yijie/code/spark/network/common/src/main/java/org/apache/spark/network/protocol/ChunkFetchFailure.java:56: error: [EqualsHashCode] Classes that override equals should also override hashCode.
  public boolean equals(Object other) {
                 ^
    (see http://errorprone.info/bugpattern/EqualsHashCode)

**Comment by yijieshen** [comment]
Also note the `[UnnecessaryStaticImport] Using static imports for types is unnecessary` conflict with a scaladoc walk around: [SPARK-7726](https://github.com/apache/spark/pull/6260/files) (`UnnecessaryStaticImport ` is also a warning by default.)

/Users/yijie/code/spark/network/shuffle/src/main/java/org/apache/spark/network/shuffle/protocol/OpenBlocks.java:28: error: [UnnecessaryStaticImport] Using static imports for types is unnecessary
import static org.apache.spark.network.shuffle.protocol.BlockTransferMessage.Type;
^
    (see http://errorprone.info/bugpattern/UnnecessaryStaticImport)
  Did you mean 'import org.apache.spark.network.shuffle.protocol.BlockTransferMessage.Type;'?
/Users/yijie/code/spark/network/shuffle/src/main/java/org/apache/spark/network/shuffle/protocol/StreamHandle.java:24: error: [UnnecessaryStaticImport] Using static imports for types is unnecessary
import static org.apache.spark.network.shuffle.protocol.BlockTransferMessage.Type;
^
    (see http://errorprone.info/bugpattern/UnnecessaryStaticImport)
  Did you mean 'import org.apache.spark.network.shuffle.protocol.BlockTransferMessage.Type;'?
/Users/yijie/code/spark/network/shuffle/src/main/java/org/apache/spark/network/shuffle/protocol/RegisterExecutor.java:26: error: [UnnecessaryStaticImport] Using static imports for types is unnecessary
import static org.apache.spark.network.shuffle.protocol.BlockTransferMessage.Type;
^
    (see http://errorprone.info/bugpattern/UnnecessaryStaticImport)
  Did you mean 'import org.apache.spark.network.shuffle.protocol.BlockTransferMessage.Type;'?
/Users/yijie/code/spark/network/shuffle/src/main/java/org/apache/spark/network/shuffle/protocol/UploadBlock.java:28: error: [UnnecessaryStaticImport] Using static imports for types is unnecessary
import static org.apache.spark.network.shuffle.protocol.BlockTransferMessage.Type;
^
    (see http://errorprone.info/bugpattern/UnnecessaryStaticImport)
  Did you mean 'import org.apache.spark.network.shuffle.protocol.BlockTransferMessage.Type;'?
4 errors

**Comment by yijieshen** [comment]
/Users/yijie/code/spark/sql/catalyst/src/main/scala/org/apache/spark/sql/BaseRow.java:143: error: [TypeParameterUnusedInFormals] Declaring a type parameter that is only used in the return type is a misuse of generics: operations on the type parameter are unchecked, it hides unsafe casts at invocations of the method, and it interacts badly with method overload resolution
  public <T> T getAs(int i) {
               ^
    (see http://errorprone.info/bugpattern/TypeParameterUnusedInFormals)
  Did you mean 'public Object getAs(int i) {'?
/Users/yijie/code/spark/sql/catalyst/src/main/scala/org/apache/spark/sql/BaseRow.java:148: error: [TypeParameterUnusedInFormals] Declaring a type parameter that is only used in the return type is a misuse of generics: operations on the type parameter are unchecked, it hides unsafe casts at invocations of the method, and it interacts badly with method overload resolution
  public <T> T getAs(String fieldName) {
               ^
    (see http://errorprone.info/bugpattern/TypeParameterUnusedInFormals)
  Did you mean 'public Object getAs(String fieldName) {'?
2 errors

**Comment by yijieshen** [comment]
Perhaps it's time to reconsider whether we should level up all the `warnings` to `error` ?

**Comment by yijieshen** [comment]
retest this please.

**Comment by yijieshen** [comment]
Jenkins, test this please.

**Comment by andrewor14** [comment]
retest this please

**Comment by andrewor14** [comment]
@yijieshen note that this doesn't merge cleanly with master

**Comment by yijieshen** [comment]
@andrewor14, thanks for retesting this. I would merge this ASAP.

**Comment by yijieshen** [comment]
@andrewor14, the latest commit merge cleanly with master now, mind reviewing it as well as triggering the test?

**Comment by rxin** [comment]
Jenkins, retest this please.

**Comment by rxin** [comment]
cc @JoshRosen will this substantially increase build time due to maven?

**Comment by yijieshen** [comment]
Run this check would take about ten minutes on my laptop

**Comment by andrewor14** [comment]
The last test seemed to take only `152m`, which is around the average.

**Comment by JoshRosen** [comment]
I've thought about this a bit and I'm not sure that ErrorProne offers us a huge win unless it's run as part of every build. I think the advantage of ErrorProne over FindBugs is that it's fast, returns few false-positives, and integrates into everyday builds. I think that we lose a lot of these advantages if we guard its use behind a separate Jenkins trigger phrase or build profile / flag; if we have to manually trigger it, then I feel that we might as well run FindBugs instead.

I don't think that we can necessarily justify adding 10 minutes of build time to the pull request builder in order to run these tests, since the cost/benefit ratio there seems low.  _However_, I'd be interested to see if there's a way to compile _just_ the Java sources with this, but that might not be possible given the Scala <-> Java interdependencies in our code base.

In light of all of this, I think that we might want to put this on hold for now and revisit later once we have more flexible Jenkins triggering logic or the ability to run the pull request builder tests using Maven.

Thanks for working on this and sorry that we're probably not going to be able to pull it in.

**Comment by yijieshen** [comment]
Ok, I will close it.

**Pull request closed without merging**